collector/timex.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build !notimex
+
+package collector
+
+// #include <sys/timex.h>
+import "C"
+
+import (
+	"fmt"
+	"syscall"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// The system clock is not synchronized to a reliable server.
+	timeError = C.TIME_ERROR
+)
+
+type timexCollector struct {
+	offset, freq, maxerror, esterror, status, constant, precision, tolerance, tick, ppsfreq, jitter, shift, stabil, jitcnt, calcnt, errcnt, stbcnt, tai, syncStatus typedDesc
+}
+
+func init() {
+	Factories["timex"] = NewTimexCollector
+}
+
+// NewTimexCollector returns a new Collector exposing adjtime(3) stats.
+func NewTimexCollector() (Collector, error) {
+	const subsystem = "timex"
+
+	return &timexCollector{
+		offset: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "offset_seconds"),
+			"Time offset in between local system and reference clock.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		freq: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "freq"),
```

Comment:
We prefer longer more descriptive metric names when possible.  The documentation for this seems to be:
```
long freq; /* Frequency offset, in units of 2^-16 ppm (parts per million, see NOTES below) */
```

Perhaps `frequency_offset_ppm` as a metric name?

Reply:
I think you refer local clock frequency adjustment. I changed that to be 'frequency_adjustment' and in same go renamed many other measurement names. Lot of them got pps (pulses  per second) prefix, and they are useful for people who run clock servers.  I reckon for 99.9999% these values are pointless. This is discussed below more in detail.